Cross chain execution using signed UserOps (#1)
magmo/Bridge-Wallet: docs/cross-chain-payments-and-execution.md

## Cross chain payments and execution

> This document suggests an approach of using `UserOp`s to perform cross-chain payments and execution. At a very high level this approach can be thought of as a cross-chain state channel, where we have some state, that when fully signed can be submitted to an adjudicator contract to claim the funds based on the outcome of that state.

Reviewer: ```suggestion
This document suggests an approach of using `UserOp`s to perform cross-chain payments and execution. At a very high level this approach can be thought of as a cross-chain state channel, where we have some state, that when fully signed can be submitted to one of many adjudicator contracts (each on a different chain) to claim the funds on that chain based on the outcome of that state.
```

However instead of implementing the state channel framework ourselves, we're making use of some of ERC 4337 infrastructure. Instead of signing a state, participants sign a **UserOp** that contains the state information. Instead of having a specific "adjudicator" contract, we have the **entrypoint** and the **BridgeWallet SCW** to handle adjudicating funds. Instead of making a on-chain call to challenge on the adjudicator, we can submit the `UserOp` to an entrypoint to trigger a challenge.

Since we can specify the UserOp calldata, we can include whatever additional state information we want in that calldata. That lets us embed payment or execution information for different chains in one UserOp. When submitted to chain the UserOp will behave depending on the chain Id and the embedded payment or execution information.

# Payments

> We embed payment information in the UserOp using a `PaymentChainInfo`

Reviewer 2: Could the structs be renamed to `PaymentInfo` and `ExecuteInfo` for simplicity? Or does the use of `Chain` in those names help clarify what they are?

Author: I think the use of `Chain` helps clarify that info in these structs is specific to a given chain, and should only be honoured on that chain. I don't feel too strongly about it though so I'm happy to change it if others feel differently.

> ```sol

Reviewer: Syntax highlighting doesn't seem to work here, have you tried `solidity`?

/// Contains information to execute a payment
struct PaymentChainInfo {
  uint chainId;
  address entrypoint;
  State paymentState;
}

```

> Whenever a fully signed UserOp containing a `PaymentChainInfo` is submitted to chain, it forces a challenge on that chain. This means that once you have a fully signed UserOp you know you can always use it to get your funds via challenge. In the happy path, we would expect participants to just exchange signed states afterwards, so they can discard the UserOp.

Reviewer: ```suggestion
Whenever a fully signed UserOp containing a `PaymentChainInfo` is submitted to chain, it forces a challenge on that chain. This means that once you have a fully signed UserOp you know you can always use it to get your funds via challenge. In the happy path, the `UserOp` need never be submitted on chain -- we would expect participants to gather signatures on the `UserOp` off chain, and then to progress their off chain state to absorb the effects of the `UserOp`. They can then discard the UserOp.
```

Reviewer 2: Is it safe to discard the signed `UserOp`? If a user needs to initiate a challenge, can they just submit a signed `State` or would they need to submit the entire signed `UserOp`.

Author: Users can always submit the signed state directly to initiate a challenge, so it should be safe to discard the UserOp after everyone signs the state.

Reviewer 2: Ok great. It might be nice to include some details indicating users can challenge using just the signed state (and maybe how that is done) and that is why it is safe to discard the signed `UserOp`

## Cross-chain Payment Example

> Let's say we have Alice, Bob,and Irene. Alice and Irene have a BridgeWallet on chain A, Bob and Irene have a BridgeWallet on Chain B. Alice and Irene have both signed a state with a balance of `[A:5,I:5]`, and bob and Irene have likewise have a signed state with a balance of`[B:5,I:5]`. Let's say Alice wants to pay Bob.

Reviewer: ```suggestion
Let's say we have Alice, Bob,and Irene. Alice has a BridgeWallet on chain A with Irene as the intermediary and Bob has a BridgeWallet on Chain B also with Irene as the intermediary. Alice and Irene have both signed a state with a balance of `[A:5,I:5]`, and bob and Irene have likewise have a signed state with a balance of`[B:5,I:5]`. Let's say Alice wants to pay Bob.
```

1. Alice creates two new unsigned states `[A:4,I:6]` and `[B:6,I:4]`

> 2. Alice includes them in a UserOp(wrapped in a `PaymentChainInfo`), and signs the UserOp. She sends this to Bob and Irene.

Reviewer 2: Wouldn't the wrapping be the other way around? The `PaymentChainInfo` is wrapped in a `UserOp`

Author: I think this is just awkward wording, I meant that the state is wrapped in a `PaymentChainInfo` (which is wrapped in the `UserOp`. I'll tweak the wording.

> 3. Bob and Irene validate the UserOp and also sign it, and broadcast it.

Reviewer: ```suggestion
3. Bob and Irene validate the `UserOp` and also sign it, and broadcast it.
```

Reviewer: Are you replacing the hashlock mechanism with signatures, or just adding to it?

Reviewer 2: What do you mean by "broadcast it" in this context? Where are Bob and Irene sending the signed `UserOp`?

Author: >Are you replacing the hashlock mechanism with signatures, or just adding to it?

In this example we are replacing the hashlock mechanism, since we're making a cross-chain payment that doesn't use HTLCs.

Author: > What do you mean by "broadcast it" in this context? Where are Bob and Irene sending the signed `UserOp`?

Broadcast in this case just means send it to the other parties. Bob is broadcasting the state to Alice and Irene, Irene is sending to Alice and Bob.

4. At this point the UserOp is fully signed and can be used to force a challenge on either chain, by submitting the signed UserOp to the entrypoint on either chain.

# Cross-chain Execution

We can also extend this idea to cross chain execution. Instead of embedding a payment state in the `UserOp` we embed information to make an on-chain function call. If the `UserOp` is submitted to the `entrypoint` the function will be executed **if the chainId matches the chain**.

We embed this information using a `ExecuteChainInfo`

```sol
/// Contains information to execute a function call
struct ExecuteChainInfo {
  uint chainId;
  address entrypoint;
  address dest;
  uint value;
  bytes callData;
  address owner;
  address intermediary;
}
```

> A UserOp can contain multiple `ExecuteChainInfo`s, meaning you can have "atomic" cross chain execution (once the UserOp is fully signed, you're guaranteed you can trigger the execution on either chain). A UserOp can also contain `PaymentChainInfo`s and `ExecuteChainInfos` allowing to pay for cross-chain execution.

Reviewer: Can you expand on the scare quotes around "atomic"?

Author: I'm not sure if atomic is the correct word, but I'm trying to describe that once the UserOp is fully signed, we can be guaranteed that the execution can be triggered on either chain.

> A UserOp can contain multiple `ExecuteChainInfo`s, meaning you can have "atomic" cross chain execution (once the UserOp is fully signed, you're guaranteed you can trigger the execution on either chain). A UserOp can also contain `PaymentChainInfo`s and `ExecuteChainInfos` allowing to pay for cross-chain execution.

Reviewer: ```suggestion
A UserOp can contain multiple `ExecuteChainInfo`s, meaning you can have "atomic" cross chain execution (once the UserOp is fully signed, you're guaranteed you can trigger the execution on either chain). A UserOp can also contain `PaymentChainInfo`s and `ExecuteChainInfos` allowing for paid cross-chain execution.
```

# Multihop

> Since a UserOp can contain multiple `PaymentChainInfo`s and `ExecuteChainInfos`, this approach can be used for multi-hop execution or payments. We just require the `owner` and `intermediary` of every BridgeWallet involved to sign the UserOp.

Reviewer: When we go beyond a single hop, we have "intermediary-intermediary" channels. Is this a bit awkward with the "owner" / "intermediary" roles of the bridge wallet? Will one of the parties just arbitrarily take the role of "owner"?

Author: I think right now it might be a bit awkward, as once intermediary would play the role of the `owner` and one intermediary would play the role of `intermediary` in the BridgeWallet. 

To solve this we could implement a specific Intermediary<->Intermediary mode (or separate contract?) for the BridgeWallet that handles the "intermediary-intermediary" use case.

# Replay Attacks

It's important that we're not vulnerable to replay attacks, where a `UserOp` is submitted again using a different entrypoint or chain.

> If we're working on one chain, it's fairly easy to prevent replay attacks. The `userOpHash` [provided by the Entrypoint](https://github.com/magmo/Bridge-Wallet/blob/ad6d24fa2435f449751d1b61e24d12faff1f83a9/contracts/core/EntryPoint.sol#L298) to `validateUserOp` is hashed against the current chain and entrypoint. This means that if the UserOp is run against a different chain or entrypoint, there will be a different `userOpHash` causing all the signature checks to fail.

Reviewer: This feels a little suboptimal. I suppose the entrypoint has not been designed with cross chain applications in mind, meaning that we have to make multiple hashes and a signature on each one (and as you say validate all of those combinations). 

A "cross chain entrypoint" would ease this, but would be a departure from 4337 norms (I guess?).

Author: With [this change](https://github.com/magmo/Bridge-Wallet/pull/1/commits/c4829f5763967725576751055c721a6dcf98e949) I don't think it's too bad. Now instead of generating a unique hash for each chain, all participants just sign the hash based on the first chain. This means we can generate one hash and check all the signatures against that.

I think it's safe to have everyone sign the same `UserOpHash` since we perform a separate check on the chainId and entrypoint based on the calldata in the `UserOp`. If you try submitting the UserOp, it will only work if the chain and entrypoint match a `ExecuteChainInfo` or `PaymentChainInfo`.

So the cost of  `validateCrossChain` would be generating one hash, and then checking 2 signatures for every `BridgeWallet` involved. IE: If we have a multi-hop UserOp with Alice,Irene,Iris,Bob we'd expect the `UserOp` to have the following signatures: `[AliceSig,IreneSig,IreneSig,IrisSig,IrisSig,BobSig]`.  We'd generate one hash and check the6 signatures against that hash.

 It would also be an easy optimization to omit or skip duplicate signatures so we could get the cost down to generating one hash and checking 1 signature per participant.

> With cross-chain execution and payments we need a slightly more complicated check. To prevent replay attacks we ensure that one of the `ExecuteChainInfo` or `PaymentChainInfo` matches our chainId and entrypoint. This ensures that the `UserOp` will only be handled by the chain/entrypoint specified by the `ExecuteChainInfo/PaymentChainInfo`.

Reviewer 2: ```suggestion
With cross-chain execution and payments we need a slightly more complicated check. To prevent replay attacks we ensure that at least one of the `ExecuteChainInfo` or `PaymentChainInfo` contains the network's chainId and entrypoint. This ensures that the `UserOp` will only be handled by the chain/entrypoint specified by the `ExecuteChainInfo/PaymentChainInfo`.
```

# Signatures

We expect the UserOp to be signed by the `owner` and `intermediary` of every BridgeWallet involved. So based on the example above with Alice,Irene,Bob, we'd expect the signatures `[AliceSigChainA,IreneSigChainA,IreneSigChainB,BobSigChainB]`.

When validating signatures we iterate through the `ExecuteChainInfo` and `PaymentChainInfo` and check the signatures against the hash generated using the chain id/entrypoint from the ChainInfo. It's important that we validate the signature for other chains, because that guarantees the UserOp is "atomic". Otherwise a partially signed UserOp could be redeemed on one chain, while not being redeemable on the other chain.

# Nonces

> We need to be careful with how we use nonces. The `owner` of a BridgeWallet can always submit a UserOp to [trigger a](https://github.com/magmo/Bridge-Wallet/blob/66dbb9c41ea8830218265b4def76824320df6bca/contracts/SCBridgeWallet.sol#L207) `Challenge` or `Reclaim` call. So if we have a UserOp signed by everyone with some `nonce`, the `owner` could submit a UserOp just signed by them burning the nonce, and preventing the UserOp signed by everyone from being handled.

Reviewer: Great job anticipating this 👏
